contracts/abstract/JBPayoutRedemptionPaymentTerminal3_1_1.sol

```diff
@@ -0,0 +1,1803 @@
+
+      // Burn the project tokens.
+      if (_tokenCount > 0)
+        IJBController(directory.controllerOf(_projectId)).burnTokensOf(
+          _holder,
+          _projectId,
+          _tokenCount,
+          '',
+          false
+        );
+
+      // If delegate allocations were specified by the data source, fulfill them.
+      if (_delegateAllocations.length != 0) {
+        // Keep a reference to the token amount being forwarded to the delegate.
+        JBTokenAmount memory _forwardedAmount = JBTokenAmount(token, 0, decimals, currency);
+
+        JBDidRedeemData3_1_1 memory _data = JBDidRedeemData3_1_1(
+          _holder,
+          _projectId,
+          _fundingCycle.configuration,
+          _tokenCount,
+          JBTokenAmount(token, reclaimAmount, decimals, currency),
+          _forwardedAmount,
+          _beneficiary,
+          _memo,
+          bytes(''),
+          _metadata
+        );
+
+        for (uint256 _i; _i < _delegateAllocations.length; ) {
+          // Get a reference to the delegate being iterated on.
+          JBRedemptionDelegateAllocation3_1_1 memory _delegateAllocation = _delegateAllocations[_i];
+
+          // Get the fee for the delegated amount.
+          uint256 _delegatedAmountFee = _feeAmount(_delegateAllocation.amount, fee, _feeDiscount);
+
+          // Add the delegated amount to the amount eligible for having a fee taken.
+          if (_delegatedAmountFee != 0) {
+            _feeEligibleDistributionAmount += _delegateAllocation.amount;
+            _delegateAllocation.amount -= _delegatedAmountFee;
+          }
+
+          // Trigger any inherited pre-transfer logic.
+          _beforeTransferTo(address(_delegateAllocation.delegate), _delegateAllocation.amount);
+
+          // Keep track of the msg.value to use in the delegate call
+          uint256 _payableValue;
+
+          // If this terminal's token is ETH, send it in msg.value.
+          if (token == JBTokens.ETH) _payableValue = _delegateAllocation.amount;
+
+          // Pass the correct token forwardedAmount to the delegate
+          _data.forwardedAmount.value = _delegateAllocation.amount;
+
+          // Pass the correct metadata from the data source.
+          _data.dataSourceMetadata = _delegateAllocation.metadata;
+
+          _delegateAllocation.delegate.didRedeem{value: _payableValue}(_data);
+
+          emit DelegateDidRedeem(
+            _delegateAllocation.delegate,
+            _data,
+            _delegateAllocation.amount,
+            _delegatedAmountFee,
+            msg.sender
+          );
+
+          unchecked {
+            ++_i;
+          }
+        }
+      }
+    }
+
+    // Keep a reference to the fee taken from the reclaim amount.
+    uint256 _reclaimAmountFee;
+
+    // Send the reclaimed funds to the beneficiary.
+    if (reclaimAmount != 0) {
+      // Get the fee for the reclaimed amount.
+      _reclaimAmountFee = _feeAmount(reclaimAmount, fee, _feeDiscount);
```

Comment:
fee is potentially sloaded twice (ln 836) + if we fix by having it on stack here, why not testing fee == 0 and shortcut?